### DISCUSSION TITLE
Carousel not showing base64 images on slide when loaded dynamically, but renders thumbnails

### DISCUSSION THREAD

**Discussion by iAmNikola** [quasarframework/quasar#14262] [General - Components / Directives / etc] [answered]
Hi! Thank You for taking the time to read through this :) 
If something bellow needs further explanation, please let me know.

I'm using Quasars carousel component (I changed the look of it with css) to show images that are in base64 format. When the images are in the component's file, as in, the base64 string literals are in the file, images are displayed in both the thumbnail and slide. That's great and it works as intended both when the list of images is populated statically and dynamically. 

The problem arises when I try to use those same base64 strings when fetched from the server. Everything is correct in terms of setting img-src, name and key for `q-carousel-slide`, but displaying those images on the slide just doesn't work. The thumbnail is loaded but it doesn't place the image on the `q-carousel__slide` div in DOM (shown bellow).

![Screenshot from 2022-08-25 10-36-41](https://user-images.githubusercontent.com/58849671/186617909-20df4bc8-7d97-4940-933e-b97a9111203c.png)
*Carousel selected on a base64 string that is a string literal in component's file*

![Screenshot from 2022-08-25 10-37-08](https://user-images.githubusercontent.com/58849671/186618605-7ab583ab-d48a-4ee4-99bf-713d44828c02.png)
*Carousel selected on a base64 string that was fetched from the server. Missing `style="background-image: url(....)"`*

I have no clue what makes it behave this way.
Things I do know:

- v-model is correct
- images fetched from server are correct (they render in the thumbnail)
- mounting the carousel after images are loaded doesn't make a difference

Code behind implementation:
```html
<q-carousel
    class="img-slides"
    animated
    arrows
    thumbnails
    infinite
    control-color="primary"
    v-model="selected_img"
    >
    <q-carousel-slide v-for="(img, i) in images" :name="i+1" :img-src="img.data" :key="img.id"/>
</q-carousel>
```

```js
const images = ref([]);
const selected_img = ref(1);
.
.
.
// images are fetched from a server and they are in string base64 format
for (let [key, el] of Object.entries(result['result'])) {
    images.value = [...images.value, {data: `data:image/png;base64,${el}`, id: `picture_${key}`}];
}
```

**Comment by MilosPaunovic** [comment]
Hi Nikola,

There does not seem to be an issue with `base64` images.
https://codepen.io/milospaunovic/pen/mdxYeMZ?editors=1010

Could you provide a Codepen reproduction of the issue?

**Reply by iAmNikola** [reply]
Hi Milos,

I have no issue displaying base64 images when they are already in file like you have provided in your codepen. The problem is when I try to display base64 images fetched from the server in the carousel. As you saw from pictures provided above, images fetched from the server render in thumbnail, but don't get placed in carousel's slide.

Hope I've made things clearer.

> Could you provide a Codepen reproduction of the issue?

The server that images are being fetched from is local so I don't know how I could codepenify that :D

**Reply by MilosPaunovic** [reply]
Could you create a fake API call, using `setTimeout` which will return you data in the same format as the one in your local example?

**Reply by iAmNikola** [reply]
I made a codepen with `setTimeout`. It works as expected.
https://codepen.io/iamnikola-the-bold/pen/MWVdKaG
It's still not the same as my case. The images are written in string literal in file when it's run. My issue happens when the images are fetched from the server (they are fetched correctly because they render in thumbnails).

You are probably as confused as I am. I've been trying to figure out what the problem with this is for 2 days :dizzy_face: 
Thanks for the your assistance btw :)

**Reply by MilosPaunovic** [reply]
Fetching data from server and acquiring it through an asynchronous `setTimeout` should be the same in regards to Quasar. What that is suggesting is that the next reasonable step would be to create some kind of minimal code reproduction, if not Codepen than a repository. Try to find a free API which will return you the `base64` image for testing purposes.

**Reply by iAmNikola** [reply]
I agree. That't why this is so confusing to me. If it hadn't rendered thumbnails I would think something is wrong with the server sending the images. but they are sent correctly.

I made a codepen which fetches images using [picsum.photos](picsum.photos) API, converts them to base64 and adds them to the images list.
https://codepen.io/iamnikola-the-bold/pen/MWVdKaG (write and delete a character in script if it doesn't run automatically)
It works! That's great, but I'm left with more questions then answers now...

The fact that images from my server render in thumbnail but not the slide is driving me insane :D

**Reply by MilosPaunovic** [reply]
Is it possible that you don't set proper value to selected image property, meaning you set an ID which does not exist in the images array?

**Reply by iAmNikola** [reply]
Checked it again, it's correct

**Comment by iAmNikola** [comment, marked as answer]
I figured out what was the issue. If you are using a library to convert images to `base64` check if it adds newline characters ("**\n**") to the result. This was the root cause of my issue. After removing all newline characters it displays images on slide.

The reason for this behavior for the curious: 
`<img>` tag ignores newline characters so it renders the image. HTML `style` property in tags, inline CSS, doesn't and it fails to read the image. Therefore, leaving the `<div class="q-carousel__slide">` without a style property.
